### PR TITLE
Add client statistics

### PR DESCRIPTION
**Pull request by roclark**
Phase 1 of fixing #69.

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

### CODE REVIEW

**Comment by roclark** [comment]
I feel pretty good about the types but definitely want the names looked at closely.

File: sources/procfs.go [hunk 1]
```diff
@@ -296,6 +296,44 @@ func (s *lustreSource) generateMDSMetricTemplates() error {
...
+	metricMap := map[string][]lustreHelpStruct{
+		"llite": {
+			{"blocksize", "blocksize_bytes", "Filesystem block size in bytes", s.gaugeMetric},
+			{"checksum_pages", "checksum_pages", "Returns '1' if data checksumming is enabled for the client", s.gaugeMetric},
+			{"default_easize", "default_easize_bytes", "Default Extended Attribute (EA) size in bytes", s.gaugeMetric},
+			{"filesfree", "inodes_free", "The number of inodes (objects) available", s.gaugeMetric},
+			{"filestotal", "inodes_maximum", "The maximum number of inodes (objects) the filesystem can hold", s.gaugeMetric},
+			{"kbytesavail", "kilobytes_available", "Number of kilobytes readily available in the pool", s.gaugeMetric},
+			{"kbytesfree", "kilobytes_free", "Number of kilobytes allocated to the pool", s.gaugeMetric},
+			{"kbytestotal", "kilobytes_capacity", "Capacity of the pool in kilobytes", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procfs.go]: I've got myself rethinking these metrics now (all the kilobytes_* ones). Should they actually be *_kilobytes instead?

File: sources/procfs.go [hunk 2]
```diff
@@ -296,6 +296,44 @@ func (s *lustreSource) generateMDSMetricTemplates() error {
...
+			{"lazystatfs", "lazystatfs", "Returns '1' if lazystatfs (a non-blocking alternative to statfs) is enabled for the client", s.gaugeMetric},
+			{"max_easize", "max_easize", "Maximum Extended Attribute (EA) size in bytes", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procfs.go]: max_easize_bytes

File: sources/procfs.go [hunk 3]
```diff
@@ -296,6 +296,44 @@ func (s *lustreSource) generateMDSMetricTemplates() error {
...
+			{"statahead_max", "statahead_max", "Maximum window size for statahead", s.gaugeMetric},
+			{"stats", "read_samples_total", readSamplesHelp, s.counterMetric},
+			{"stats", "read_minimum_size_bytes", readMinimumHelp, s.gaugeMetric},
+			{"stats", "read_maximum_size_bytes", readMaximumHelp, s.gaugeMetric},
+			{"stats", "read_bytes_total", readTotalHelp, s.counterMetric},
+			{"stats", "write_samples_total", writeSamplesHelp, s.counterMetric},
+			{"stats", "write_minimum_size_bytes", writeMaximumHelp, s.gaugeMetric},
+			{"stats", "write_maximum_size_bytes", writeMaximumHelp, s.gaugeMetric},
+			{"stats", "write_bytes_total", writeTotalHelp, s.counterMetric},
+			{"xattr_cache", "xattr_cache", "Returns '1' if extended attribute cache is enabled", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procfs.go]: maybe rename xattr_cache_enabled?

Comment (roclark) [inline sources/procfs.go]: Works for me!

File: sources/procfs.go [hunk 4]
```diff
@@ -296,6 +296,44 @@ func (s *lustreSource) generateMDSMetricTemplates() error {
...
+			{"kbytesavail", "kilobytes_available", "Number of kilobytes readily available in the pool", s.gaugeMetric},
+			{"kbytesfree", "kilobytes_free", "Number of kilobytes allocated to the pool", s.gaugeMetric},
+			{"kbytestotal", "kilobytes_capacity", "Capacity of the pool in kilobytes", s.gaugeMetric},
+			{"lazystatfs", "lazystatfs", "Returns '1' if lazystatfs (a non-blocking alternative to statfs) is enabled for the client", s.gaugeMetric},
+			{"max_easize", "max_easize", "Maximum Extended Attribute (EA) size in bytes", s.gaugeMetric},
+			{"max_read_ahead_mb", "max_read_ahead_megabytes", "Maximum number of megabytes to read ahead", s.gaugeMetric},
+			{"max_read_ahead_per_file_mb", "max_read_ahead_per_file_megabytes", "Maximum number of megabytes per file to read ahead", s.gaugeMetric},
+			{"max_read_ahead_whole_mb", "max_read_ahead_whole_megabytes", "Maximum file size in megabytes for a file to be read in its entirety", s.gaugeMetric},
+			{"statahead_agl", "statahead_agl", "Returns '1' if the Asynchronous Glimpse Lock (AGL) for statahead is enabled", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procfs.go]: statahead_agl_enabled? Or maybe even worth expanding out to statahead_asynchronous_glimpse_lock_enabled...though that is super long...

**Comment by roclark** [comment]
Added `_enabled` to the end of all binary enabled/disabled metrics and changed `max_easize` to `max_easize_bytes`.

File: sources/procfs.go (hunk 1 above)
Comment (roclark) [inline sources/procfs.go]: That does make sense, especially from a consistency standpoint. Echoing an offline discussion, we want to save this for another PR, correct?

Comment (joehandzik) [inline sources/procfs.go]: Yes, correct.

File: sources/procfs.go [hunk 5]
```diff
@@ -296,6 +296,44 @@ func (s *lustreSource) generateMDSMetricTemplates() error {
...
+			{"blocksize", "blocksize_bytes", "Filesystem block size in bytes", s.gaugeMetric},
+			{"checksum_pages", "checksum_pages_enabled", "Returns '1' if data checksumming is enabled for the client", s.gaugeMetric},
+			{"default_easize", "default_easize_bytes", "Default Extended Attribute (EA) size in bytes", s.gaugeMetric},
+			{"filesfree", "inodes_free", "The number of inodes (objects) available", s.gaugeMetric},
+			{"filestotal", "inodes_maximum", "The maximum number of inodes (objects) the filesystem can hold", s.gaugeMetric},
+			{"kbytesavail", "kilobytes_available", "Number of kilobytes readily available in the pool", s.gaugeMetric},
+			{"kbytesfree", "kilobytes_free", "Number of kilobytes allocated to the pool", s.gaugeMetric},
+			{"kbytestotal", "kilobytes_capacity", "Capacity of the pool in kilobytes", s.gaugeMetric},
+			{"lazystatfs", "lazystatfs_enabled", "Returns '1' if lazystatfs (a non-blocking alternative to statfs) is enabled for the client", s.gaugeMetric},
+			{"max_easize", "max_easize_bytes", "Maximum Extended Attribute (EA) size in bytes", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procfs.go]: Worth adding an underscore to make it max_ea_size_bytes?

Comment (roclark) [inline sources/procfs.go]: Done!

File: sources/procfs.go [hunk 6]
```diff
@@ -296,6 +296,44 @@ func (s *lustreSource) generateMDSMetricTemplates() error {
...
+			{"kbytesfree", "kilobytes_free", "Number of kilobytes allocated to the pool", s.gaugeMetric},
+			{"kbytestotal", "kilobytes_capacity", "Capacity of the pool in kilobytes", s.gaugeMetric},
+			{"lazystatfs", "lazystatfs_enabled", "Returns '1' if lazystatfs (a non-blocking alternative to statfs) is enabled for the client", s.gaugeMetric},
+			{"max_easize", "max_ea_size_bytes", "Maximum Extended Attribute (EA) size in bytes", s.gaugeMetric},
+			{"max_read_ahead_mb", "max_read_ahead_megabytes", "Maximum number of megabytes to read ahead", s.gaugeMetric},
+			{"max_read_ahead_per_file_mb", "max_read_ahead_per_file_megabytes", "Maximum number of megabytes per file to read ahead", s.gaugeMetric},
+			{"max_read_ahead_whole_mb", "max_read_ahead_whole_megabytes", "Maximum file size in megabytes for a file to be read in its entirety", s.gaugeMetric},
+			{"statahead_agl", "statahead_agl_enabled", "Returns '1' if the Asynchronous Glimpse Lock (AGL) for statahead is enabled", s.gaugeMetric},
+			{"statahead_max", "statahead_max", "Maximum window size for statahead", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procfs.go]: Should we make this statahead_maximum to comply with the precedent we just started with our other _maximum statistic?

File: sources/procfs.go [hunk 7]
```diff
@@ -296,6 +296,44 @@ func (s *lustreSource) generateMDSMetricTemplates() error {
...
+			{"filesfree", "inodes_free", "The number of inodes (objects) available", s.gaugeMetric},
+			{"filestotal", "inodes_maximum", "The maximum number of inodes (objects) the filesystem can hold", s.gaugeMetric},
+			{"kbytesavail", "kilobytes_available", "Number of kilobytes readily available in the pool", s.gaugeMetric},
+			{"kbytesfree", "kilobytes_free", "Number of kilobytes allocated to the pool", s.gaugeMetric},
+			{"kbytestotal", "kilobytes_capacity", "Capacity of the pool in kilobytes", s.gaugeMetric},
+			{"lazystatfs", "lazystatfs_enabled", "Returns '1' if lazystatfs (a non-blocking alternative to statfs) is enabled for the client", s.gaugeMetric},
+			{"max_easize", "max_ea_size_bytes", "Maximum Extended Attribute (EA) size in bytes", s.gaugeMetric},
+			{"max_read_ahead_mb", "max_read_ahead_megabytes", "Maximum number of megabytes to read ahead", s.gaugeMetric},
+			{"max_read_ahead_per_file_mb", "max_read_ahead_per_file_megabytes", "Maximum number of megabytes per file to read ahead", s.gaugeMetric},
+			{"max_read_ahead_whole_mb", "max_read_ahead_whole_megabytes", "Maximum file size in megabytes for a file to be read in its entirety", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procfs.go]: Should these two be maximum_*?

Comment (roclark) [inline sources/procfs.go]: I'm all for it, but do you mean the four `max_` metrics? If it's just two, which ones and for what reason just for my curiosity?

Comment (joehandzik) [inline sources/procfs.go]: eh, yeah, I meant four. I edited the comment to say three, but clearly I can't count. Let's rephrase more clearly:

We should change any case of max to maximum instead. :)

Comment (roclark) [inline sources/procfs.go]: Haha, all good! Just wanted to clarify!

**Pull request merged**